Mathlib/LinearAlgebra/Semisimple.lean

```diff
@@ -121,23 +100,67 @@ lemma IsSemisimple_smul (t : K) (h : f.IsSemisimple) :
   wlog ht : t ≠ 0; · simp [not_not.mp ht]
   rwa [IsSemisimple_smul_iff ht]
 
-open UniqueFactorizationMonoid in
 theorem isSemisimple_of_squarefree_aeval_eq_zero [FiniteDimensional K M]
     {p : K[X]} (hp : Squarefree p) (hpf : aeval f p = 0) : f.IsSemisimple := by
-  classical
-  have := (Submodule.isInternal_prime_power_torsion_of_pid <|
-    AEval.isTorsion_of_finiteDimensional K M f).submodule_iSup_eq_top
-  rw [AEval.annihilator_top_eq_ker_aeval, minpoly.ker_aeval_eq_span_minpoly,
-    Ideal.submodule_span_eq, factors_eq_normalizedFactors] at this
-  refine isSemisimpleModule_of_isSemisimpleModule_submodule'
-    (fun ⟨q, hq₁⟩ ↦ Submodule.isSemisimple_torsionBy_of_irreducible <| Prime.irreducible ?_) this
-  simp only [Multiset.mem_toFinset] at hq₁
-  simp only [prime_pow_iff]
-  refine ⟨Ideal.prime_generator_of_prime (prime_of_normalized_factor q hq₁),
-    Multiset.count_eq_one_of_mem ?_ hq₁⟩
-  have hf : Ideal.span {minpoly K f} ≠ 0 := by simpa using minpoly.ne_zero_of_finite K f
-  rw [← squarefree_iff_nodup_normalizedFactors hf, Ideal.squarefree_span_singleton]
-  exact hp.squarefree_of_dvd (minpoly.dvd K f hpf)
+  rw [← RingHom.mem_ker, ← AEval.annihilator_eq_ker_aeval (M := M), mem_annihilator,
+      ← IsTorsionBy, ← isTorsionBySet_singleton_iff, isTorsionBySet_iff_is_torsion_by_span] at hpf
+  let R := K[X] ⧸ Ideal.span {p}
+  have : IsReduced R :=
+    (Ideal.isRadical_iff_quotient_reduced _).mp (isRadical_iff_span_singleton.mp hp.isRadical)
+  have : FiniteDimensional K R := (AdjoinRoot.powerBasis hp.ne_zero).finite
+  haveI : IsArtinianRing R := .of_finite K R
+  letI : Module R (AEval' f) := Module.IsTorsionBySet.module hpf
+  let e : AEval' f →ₛₗ[Ideal.Quotient.mk (Ideal.span {p})] AEval' f :=
+    { AddMonoidHom.id _ with map_smul' := fun _ _ ↦ rfl }
+  exact (e.isSemisimpleModule_iff_of_bijective bijective_id).mpr inferInstance
+
+variable [FiniteDimensional K M]
+
+/-- The minimal polynomial of a semisimple endomorphism is square free -/
+theorem IsSemisimple.minpoly_squarefree (hf : f.IsSemisimple) : Squarefree (minpoly K f) :=
+  IsRadical.squarefree (minpoly.ne_zero <| isIntegral _) <| by
+    rw [isRadical_iff_span_singleton, span_minpoly_eq_annihilator]; exact hf.annihilator_isRadical
+
+variable [PerfectField K] (comm : Commute f g) (hf : f.IsSemisimple) (hg : g.IsSemisimple)
+
+theorem isSemisimple_of_mem_adjoin {a : End K M} (ha : a ∈ Algebra.adjoin K {f, g}) :
+    a.IsSemisimple := by
+  let R := K[X] ⧸ Ideal.span {minpoly K f}
+  let S := AdjoinRoot ((minpoly K g).map <| algebraMap K R)
+  haveI : Finite K R := (AdjoinRoot.powerBasis' <| minpoly.monic <| isIntegral f).finite
```

Comment:
I had understood that in Lean4, haveI doesn't need that an instance is added. Probably have is sufficient.

Reply:
Yeah, the `I` in `haveI` stands for `inline` in Lean 4 rather than `instance`, but I habitually write `haveI` to signify instances, but maybe that's pointless. Removed the `I` now.